text/0000-action-creators.md

```diff
@@ -0,0 +1,107 @@
+- Start Date: 2018-02-09
+- RFC PR: (leave this empty)
+- React Navigation Issue: (leave this empty)
+
+# Modular Actions Creators
+
+Different routers should expose different actions. For example, it may make sense to 'pop' within a StackRouter, but an app with a drawer router should have `openDrawer` and `closeDrawer` actions available.
+
+AddNavigationHelpers forces `createNavigator` to support every action that every router supports, which prevents custom routers from exposing new action types. This is a proposal to replace `addNavigationHelpers` with a more modular set of navigation actions, defined by each router. This also allows us to deprecate the global concept of "NavigationActions" entirely.
+
+## Basic Example
+
+```
+// Only available in a Stack:
+this.props.navigation.popToTop();
+
+// Only available in a Drawer:
+this.props.navigation.openDrawer();
+```
+
+## Detailed design
+
+We will add a new function on the router, called "getActionsForRoute", which will return action creators. For example, in the StackRouter:
+
+```
+const stackRouter = {
+    getActionsForRoute: route => ({
+        popToTop: () => ({ type: 'PopToTop' }),
+        setParams: (params) => ({ type: 'SetParams', key: route.key, params }),
+        ...
+    }),
+    ...
+}
+```
+
+Inside of createNavigator, when the child navigation prop gets prepared, it will use the router to get the action creators. It will make the actions available within `childNavigation.actions`, and will also merge in parent actions provided via `this.props.navigation.actions`. That will enable the following:
+
+```
+// this.props.navigation.actions will only have `pop` within stacks:
+const action = this.props.navigation.actions.pop();
+this.props.navigation.dispatch(action);
+```
+
+As a convenience, the actions are also exposed on the navigation prop. These will call `dispatch` under the hood automatically:
+
+```
+// Only available in a Stack:
+this.props.navigation.popToTop();
+// Only available in a Drawer:
+this.props.navigation.openDrawer();
+// Only available in Tabs or Drawer:
+this.props.navigation.jumpTo();
+```
+
+This allows us to have the convenience of `addNavigationHelpers`, but each router can provide contextual helpers.
+
+
+## Action Creator Definitions
+
+Now is a good time to revisit the helpers that we have available, and what arguments they should.
+
+### Stack Router
+    - Navigate
+    - GoBack
+    - Push
```

Comment:
i wonder how this interacts https://github.com/react-navigation/react-navigation/pull/3617.

imagine we have this setup:
tabnavigator -> A: stackA, B: screenB
 
if a user wants to "push" a screen from screenB, they wouldn't have access to the `push` helper, right?

Reply:
Yeah, correct. Its still possible to construct an action to push on that stack, but the helper wouldn't be there from screen B

Reply:
That sounds like a pretty big issue with this RFC we need to solve. Besides the `push` issue there are other actions such as the DrawerNavigator's `openDrawer` helper that will pretty much always used in navigators other than the navigator itself.

Reply:
I don't quite see the issue. When do you use openDrawer from outside the drawer navigator?

Reply:
Last I checked generally the menu button to open the drawer is part of the StackNavigator's header and use's the StackNavigator's `navigation` not the DrawerNavigator's.

Reply:
You mean, if the whole drawer navigator is inside one screen of the stack navigator? It's true that wouldn't work, but I'm not sure this is one of our recommended use cases. Usually you want the drawer to be outside your stack.

Reply:
DrawerNavigator { SomeRoute: StackNavigator{} } is what I'm referring to. I haven't heard anything about menu being handled implicitly yet, so the menu button you add to the StackNavigator's header will still be using the StackNavigator's navigation, not the D.rawerNavigator's

Reply:
If you put the stack navigator within the drawer navigator like that, the navigation prop will flow down, and yes you will have navigation.openDrawer from the header configuration. I'll make sure that an example is included to demonstrate this use case.